pghoard/rohmu/object_storage/sftp.py

```diff
@@ -0,0 +1,209 @@
+                if deep:
+                    yield from self.iter_key(file_key, with_metadata=with_metadata, deep=True)
+                else:
+                    yield IterKeyItem(type=KEY_TYPE_PREFIX, value=file_key)
+            else:
+                # Don't return files if metadata file is not present; files are written in two phases and
+                # should be considered available only after also metadata has been written
+                if attr.filename + ".metadata" in files:
+                    if with_metadata:
+                        metadata = self.get_metadata_for_key(file_key)
+                    else:
+                        metadata = None
+
+                    last_modified = datetime.datetime.fromtimestamp(attr.st_mtime, tz=datetime.timezone.utc)
+                    yield IterKeyItem(
+                        type=KEY_TYPE_OBJECT,
+                        value={
+                            "name": file_key,
+                            "size": attr.st_size,
+                            "last_modified": last_modified,
+                            "metadata": metadata,
+                        }
+                    )
+                else:
+                    continue
+
+    # can't support remote copy, only remote rename
+    def copy_file(self, *, source_key, destination_key, metadata=None, **_kwargs):
+        raise NotImplementedError
+
+    def delete_key(self, key):
+        target_path = self.format_key_for_backend(key.strip("/"))
+        self.log.info("Removing path: %r", target_path)
+        try:
+            self.client.remove(target_path)
+        except FileNotFoundError as ex:
+            raise FileNotFoundFromStorageError(key) from ex
+
+    def store_file_from_memory(self, key, memstring, metadata=None, cache_control=None, mimetype=None):
+        bio = BytesIO(memstring)
+        try:
+            self.store_file_object(key=key, fd=bio, cache_control=cache_control, metadata=metadata, mimetype=mimetype)
+        except OSError as ex:
+            raise StorageError(key) from ex
+
+    def store_file_from_disk(self, key, filepath, metadata=None, multipart=None, cache_control=None, mimetype=None):
+        with open(filepath, "rb") as fh:
+            self.store_file_object(key=key, fd=fh, cache_control=cache_control, metadata=metadata, mimetype=mimetype)
+
+    # pylint: disable=unused-argument
+    def store_file_object(self, key, fd, *, cache_control=None, metadata=None, mimetype=None, upload_progress_fn=None):
+        target_path = self.format_key_for_backend(key.strip("/"))
+
+        self.log.debug("Store path: %r", target_path)
+
+        def wrapper_upload_progress_fn(bytes_written, total_bytes):  # pylint: disable=unused-argument
+            if upload_progress_fn:
+                upload_progress_fn(bytes_written)
+
+        self.client.makedirs(os.path.dirname(target_path))
+        self.client.putfo(flo=fd, remotepath=target_path, callback=wrapper_upload_progress_fn)
```

Comment:
I didn't check how this is actually implemented but storing the file should be done in somehow atomic manner so that listing files does not return files that are not yet fully uploaded. If this is automatically handled then that's fine, if not then presumably you should first upload with some temp name that is ignored from listing and then do rename once the upload is complete.

Reply:
I use the metadata file to exclude files currently being copied.  The metadata file should be small enough its written in a single block, no better way that I can see with sftp protocol